common/cuda_hip/matrix/dense_kernels.hpp.inc

```diff
@@ -33,6 +33,163 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+                block_size * block_size;
+            // which thread does it belong to?
+            const auto last_block_base_thread = last_block_base_col - base_col;
+            previous_block_nonzero =
+                bool((last_block_base_thread < 0
+                          ? nonzero_mask << -last_block_base_thread
+                          : nonzero_mask >> last_block_base_thread) &
+                     block_mask);
+        } else {
+            previous_block_nonzero = false;
+        }
+    }
+    block_count = reduce(warp, block_count,
+                         [](IndexType a, IndexType b) { return a + b; });
+    if (lane == 0) {
+        block_row_nnz[brow] = block_count;
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void convert_to_fbcsr(
+    size_type num_block_rows, size_type num_block_cols, size_type stride,
+    int block_size, const ValueType* __restrict__ source,
+    const IndexType* __restrict__ block_row_ptrs,
+    IndexType* __restrict__ block_cols, ValueType* __restrict__ blocks)
+{
+    const auto brow = thread::get_subwarp_id_flat<config::warp_size>();
+
+    if (brow >= num_block_rows) {
+        return;
+    }
+
+    const auto bs_sq = block_size * block_size;
+    const auto num_cols = num_block_cols * block_size;
+    auto warp =
+        group::tiled_partition<config::warp_size>(group::this_thread_block());
+    const auto lane = static_cast<IndexType>(warp.thread_rank());
+    auto lane_prefix_mask =
+        (config::lane_mask_type(1) << warp.thread_rank()) - 1;
+    const auto block_mask =
+        block_size >= config::warp_size
+            ? ~config::lane_mask_type{}
+            : (config::lane_mask_type{1} << block_size) - 1u;
+    bool previous_block_nonzero = false;
+    auto block_base_nz = block_row_ptrs[brow];
+    for (IndexType base_col = 0; base_col < num_cols;
+         base_col += config::warp_size) {
+        const auto col = base_col + lane;
+        const auto block_local_col = col % block_size;
+        // which is the first column in the current block?
+        const auto block_base_col = col - block_local_col;
+        // which lowest thread in the warp does this column belong to?
+        const auto block_base_thread = block_base_col - base_col;
+        // collect nonzero bitmask
+        config::lane_mask_type nonzero_mask{previous_block_nonzero ? 1u : 0u};
+        for (int local_row = 0; local_row < block_size; local_row++) {
+            const auto row = local_row + brow * block_size;
+            nonzero_mask |= warp.ballot(col < num_cols &&
+                                        is_nonzero(source[row * stride + col]));
+        }
+        const auto local_mask =
+            (block_base_thread < 0 ? nonzero_mask << -block_base_thread
+                                   : nonzero_mask >> block_base_thread) &
+            block_mask;
```

Comment:
```suggestion
            ((block_base_thread < 0) * (nonzero_mask << -block_base_thread)
                                   + (block_base_thread >= 0) * (nonzero_mask >> block_base_thread)) &
            block_mask;
```
<< negative is undefined behavior. if it never produces NaN or Inf, this should be okay to reduce the branch.
maybe use abs on block_base_thread? but I think gko::abs is also a branch